nx/source/services/ncm.c

```diff
@@ -1,1029 +1,451 @@
+#define NX_SERVICE_ASSUME_NON_DOMAIN
 #include <string.h>
+#include "service_guard.h"
+#include "runtime/hosversion.h"
 #include "services/ncm.h"
-#include "arm/atomics.h"
 
 static Service g_ncmSrv;
-static u64 g_ncmRefCnt;
 
-Result ncmInitialize(void) {
-    atomicIncrement64(&g_ncmRefCnt);
-
-    if (serviceIsActive(&g_ncmSrv))
-        return 0;
+NX_GENERATE_SERVICE_GUARD(ncm);
 
+Result _ncmInitialize(void) {
     return smGetService(&g_ncmSrv, "ncm");
 }
 
-void ncmExit(void) {
-    if (atomicDecrement64(&g_ncmRefCnt) == 0) {
-        serviceClose(&g_ncmSrv);
-    }
+void _ncmCleanup(void) {
+    serviceClose(&g_ncmSrv);
 }
 
 Service* ncmGetServiceSession(void) {
     return &g_ncmSrv;
 }
 
-Result ncmOpenContentStorage(FsStorageId storage, NcmContentStorage* out) {
-    IpcCommand c;
-    ipcInitialize(&c);
-    
-    struct {
-        u64 magic;
-        u64 cmd_id;
-        u32 storage_id;
-    } *raw;
-    
-    raw = ipcPrepareHeader(&c, sizeof(*raw));
-    
-    raw->magic = SFCI_MAGIC;
-    raw->cmd_id = 4;
-    raw->storage_id = (u32)storage;
-    
-    Result rc = serviceIpcDispatch(&g_ncmSrv);
-    if (R_SUCCEEDED(rc)) {
-        IpcParsedCommand r;
-        ipcParse(&r);
-
-        struct {
-            u64 magic;
-            u64 result;
-        } *resp = r.Raw;
-
-        rc = resp->result;
-
-        if (R_SUCCEEDED(rc)) {
-            serviceCreate(&out->s, r.Handles[0]);
-        }
-    }
-    
-    return rc;
+Result ncmCreateContentStorage(FsStorageId storage_id) {
+    return serviceDispatchIn(&g_ncmSrv, 0, storage_id);
```

Comment:
Should call a helper func instead of having multiple duplicate funcs which are identical besides the cmd_id, for example: `static Result _ncmCmdInU64(Service* srv, u32 cmd_id, u64 inval)`